port/linux/omrvmem.c

```diff
@@ -1228,6 +1228,218 @@ reserve_memory_with_mmap(struct OMRPortLibrary *portLibrary, void *address, uint
+		}
+		if (0 != (OMRPORT_VMEM_MEMORY_MODE_COMMIT & mode)) {
+			protectionFlags = get_protectionBits(mode);
+		}
+		update_vmemIdentifier(identifier, NULL, NULL, 0, 0, 0, 0, 0, NULL, -1);
+
+		void *memPtr = mmap(
+			address,
+			byteAmount,
+			protectionFlags,
+			flags,
+			fd,
+			0);
+
+		if (memPtr == MAP_FAILED) {
+			Trc_PRT_double_map_getContiguousMem_Release_Failure();
+			portLibrary->error_set_last_error_with_message(portLibrary, OMRPORT_ERROR_VMEM_OPFAILED, "Failed to map FIXED block of memory, mmap returned MAP_FAILED");
+			rc = -1;
+		} else if (memPtr != address) {
+			Trc_PRT_double_map_getContiguousMem_Release_Failure2(address, memPtr);
+			portLibrary->error_set_last_error_with_message(portLibrary, OMRPORT_ERROR_VMEM_OPFAILED, "Failed to map FIXED block of memory, mapped address differ from fixed address");
+			rc = -1;
+		}
+	} else {
+		rc = omrvmem_free_memory(portLibrary, address, byteAmount, identifier);
+	}
+
+	Trc_PRT_double_map_getContiguousMem_Release_Exit(rc);
+	return rc;
+}
+
+/**
+ *  Double maps a contiguous region of memory to discontiguous regions stored in regionAddresses[].
+ *  If preferedAddress is NULL it creates a contiguous virtual representation of memory;
+ *  otherwise, it uses preferedAddress as the contiguous region (under the assumption that
+ *  the caller reserved such contiguous region in memory).
+ *
+ *  @param OMRPortLibrary               *portLibrary        [in] The port library object
+ *  @param void                         *regionAddresses[]  [in] Addresses to be double mapped
+ *  @param uintptr_t                    regionsCount        [in] Number of regions to be double mapped
+ *  @param uintptr_t                    regionSize          [in] Size of each region
+ *  @param uintptr_t                    byteAmount          [in] Total size to allocate for contiguous block of memory
+ *  @param struct J9PortVmemIdentifier  *oldIdentifier      [in] Old Identifier containing file descriptor
+ *  @param struct J9PortVmemIdentifier  *newIdentifier      [out] New Identifier for new block of memory. The structure to be updated
+ *  @param uintptr_t                    mode                [in] Access mode
+ *  @param uintptr_t                    pageSize            [in] Constant describing page size
+ *  @param OMRMemCategory               *category           [in] Memory allocation category
+ *  @param void                         *preferedAddress    [in] Prefered address of contiguous region to be double mapped
+ * 
+ * @return pointer to contiguous region to which regions were double mapped into, NULL is returned if unsuccessful
+ */
+
+void *
+omrvmem_create_double_mapped_region(struct OMRPortLibrary *portLibrary, void* regionAddresses[], uintptr_t regionsCount, uintptr_t regionSize, uintptr_t byteAmount, struct J9PortVmemIdentifier *oldIdentifier, struct J9PortVmemIdentifier *newIdentifier, uintptr_t mode, uintptr_t pageSize, OMRMemCategory *category, void *preferedAddress)
```

Comment:
now that you are changing signature, you have a chance to remove byteAmount. you don't have to. it's just a personal preference

Reply:
Actually I'll fix this. It would probably be better to allow the caller to set the size required to be double mapped, and here we can just Assert that it is indeed what we expect:
```
Assert_PRT_true((regionsCount * regionSize) == byteAmount);
```
What do you think?